x/precisebank/keeper/burn.go

```diff
@@ -0,0 +1,177 @@
+package keeper
+
+import (
+	"errors"
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/kava-labs/kava/x/precisebank/types"
+)
+
+// BurnCoins burns coins deletes coins from the balance of the module account.
+// It will panic if the module account does not exist or is unauthorized.
+func (k Keeper) BurnCoins(ctx sdk.Context, moduleName string, amt sdk.Coins) error {
+	// Custom protection for x/precisebank, no external module should be able to
+	// affect reserves.
+	if moduleName == types.ModuleName {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "module account %s cannot be burned from", moduleName))
+	}
+
+	// Panic errors are identical to x/bank for consistency.
+	acc := k.ak.GetModuleAccount(ctx, moduleName)
+	if acc == nil {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", moduleName))
+	}
+
+	if !acc.HasPermission(authtypes.Burner) {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "module account %s does not have permissions to burn tokens", moduleName))
+	}
+
+	// Ensure the coins are valid before burning
+	if !amt.IsValid() {
+		return errorsmod.Wrap(sdkerrors.ErrInvalidCoins, amt.String())
+	}
+
+	// Get non-ExtendedCoinDenom coins
+	passthroughCoins := amt
+
+	extendedAmount := amt.AmountOf(types.ExtendedCoinDenom)
+	if extendedAmount.IsPositive() {
+		// Remove ExtendedCoinDenom from the coins as it is managed by x/precisebank
+		removeCoin := sdk.NewCoin(types.ExtendedCoinDenom, extendedAmount)
+		passthroughCoins = amt.Sub(removeCoin)
+	}
+
+	// Coins unmanaged by x/precisebank are passed through to x/bank
+	if !passthroughCoins.Empty() {
+		if err := k.bk.BurnCoins(ctx, moduleName, passthroughCoins); err != nil {
+			return err
+		}
+	}
+
+	// No more processing required if no ExtendedCoinDenom
+	if extendedAmount.IsZero() {
+		return nil
+	}
+
+	return k.burnExtendedCoin(ctx, moduleName, extendedAmount)
+}
+
+// burnExtendedCoin burns the fractional amount of the ExtendedCoinDenom from the module account.
+func (k Keeper) burnExtendedCoin(
+	ctx sdk.Context,
+	moduleName string,
+	amt sdkmath.Int,
+) error {
+	// Get the module address
+	moduleAddr := k.ak.GetModuleAddress(moduleName)
+
+	// We only need the fractional balance to burn coins, as integer burns will
+	// return errors on insufficient funds.
+	prevFractionalBalance := k.GetFractionalBalance(ctx, moduleAddr)
+
+	integerBurnAmount := amt.Quo(types.ConversionFactor())
+	fractionalBurnAmount := amt.Mod(types.ConversionFactor())
+
+	// newFractionalBalance can be negative if fractional balance is insufficient.
+	newFractionalBalance := prevFractionalBalance.Sub(fractionalBurnAmount)
+
+	// Not enough fractional balance, need to arithmetic borrow from the integer
+	// balance.
+	if newFractionalBalance.IsNegative() {
```

Comment:
Similar to the mint case, we can optimize the borrow & positive reserve case so just burn an extra coin from the account, then leave the reserve unchanged.  This will save us two bank operations, which is quite a bit of extra gas.